tc_portal_sync.py

```diff
@@ -0,0 +1,472 @@
+            self.formatter = self.format_leef
+
+    def output_event(self, event):
+        if self.flatten:
+            for flattened_event in flatten_event(event):
+                content = self.formatter(flattened_event)
+                print(content)
+        else:
+            content = self.formatter(event)
+            print(content)
+
+    def output_no_results(self):
+        pass
+
+    def format_json(self, event):
+        mapped_event = apply_mapping(self.mapping, event)
+        return json.dumps(mapped_event).encode("UTF-8")
+
+    @staticmethod
+    def format_leef(event):
+        """Format an event as QRadar / LEEF"""
+
+        syslog_header = f'<13>1 {event["actionTime"]} {hostname}'
+        leef_header = f'LEEF:2.0|TrinityCyber|PTI|1|{event.pop("id")}|xa6|'
+        fields = dict()
+
+        fields["devTime"] = event.pop("actionTime")
+        fields[
+            "devTimeFormat"
+        ] = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX"  # (e.g. 2022-04-25T00:01:19.109+00:00)
+
+        # LEEF-standard fields
+        if "source" in event:
+            fields["src"] = event.pop("source")
+        if "destination" in event:
+            fields["dst"] = event.pop("destination")
+        if "sourcePort" in event:
+            fields["srcPort"] = event.pop("sourcePort")
+        if "destinationPort" in event:
+            fields["dstPort"] = event.pop("destinationPort")
+        if "transportProtocol" in event:
+            fields["proto"] = event.pop("transportProtocol")
+
+        # Formula-related metadata
+        formula_metadata = event.pop("formula")
+        fields["tcFormulaId"] = formula_metadata["formulaId"]
+        fields["tcFormulaTitle"] = formula_metadata["title"]
+        for key, value in formula_metadata["tags"].items():
+            key = "tcFormula" + key.title().replace(" ", "")
+            fields[key] = value
+
+        # Application / protocol related data
+        for app_fields in event.pop("applicationData"):
+            for key, value in app_fields.items():
+                if value is None:
+                    continue
+                if isinstance(value, str):
+                    # Escape delimiter
+                    value = value.replace("\xa6", "\\\xa6")
+                fields[key] = value
+
+        # Add any extra information from the query
+        fields.update(event)
+
+        fields_formatted = "\xa6".join(
+            [f"{key}={value}" for key, value in fields.items()]
+        )
+        return f"{syslog_header} {leef_header}{fields_formatted}"
+
+
+class BaseBucketEventOutput:
+    def __init__(self, output_config):
+        self.flatten = output_config["flatten"]
+        self.key_base = output_config["key_base"]
+        self.flatten = output_config["flatten"]
+        self.mapping = output_config.get("field_mapping", {})
+        assert (
+            output_config.get("format", "json") == "json"
+        ), "This output only accepts JSON format"
+
+        if self.key_base:
```

Comment:
I tested the S3 upload with an empty key base.  `/filename` and `filename` are the same thing to S3.   I stripped the leading slash so the same key could be used in the directory output without writing to the root filesystem.

Reply:
Hrm; something didn't work when I messed with it last time that I put all of that logic in previously. I'm not sure using an empty path is a good idea with systemd as an option; I think it uses `/` as the `cwd` by default.

What I'm mostly trying to avoid is `//path/to/thing` because S3 sees that as a distinct key -- or `/path/to/thing/` and then appending `/filename` and getting `/path/to/thing//filename`. Also leaving empty fields should behave "sanely" (like using `/` in S3).

----

I think part of it was the explicit >>>> implicit, too. It was just easier to work with it being explicitly set. If this works for all cases with sane behavior, we can leave it. I know I spent a long time fussing with the logic on the old one to make it resilient to "uninformed user" inputs.